src/actions/genSale/purchaseDomains.ts

```diff
@@ -0,0 +1,121 @@
+import * as ethers from "ethers";
+import { getSaleStatus } from ".";
+import { GenSale } from "../../contracts/types";
+import { Claim, SaleStatus, Mintlist, Maybe, GenSaleStatus } from "../../types";
+
+const abi = ["function masterCopy() external view returns (address)"];
+
+const errorCheck = (condition: boolean, errorMessage: string) => {
+    if (condition) {
+        throw new Error(errorMessage);
+    }
+};
+
+const generateAccessList = (
+    userAddress: string,
+    gnosisSafeProxyAddress: string,
+    gnosisSafeImplAddress: string
+) => {
+    return [
+        {
+            address: gnosisSafeProxyAddress,
+            storageKeys: [
+                "0x0000000000000000000000000000000000000000000000000000000000000000",
+            ],
+        },
+        {
+            address: gnosisSafeImplAddress,
+            storageKeys: [
+                "0x0000000000000000000000000000000000000000000000000000000000000000",
+            ],
+        },
+        {
+            address: userAddress,
+            storageKeys: [],
+        },
+    ];
+};
+
+export const purchaseDomains = async (
+    count: ethers.BigNumber,
+    signer: ethers.Signer,
+    contract: GenSale,
+    mintlist: Mintlist
+): Promise<ethers.ContractTransaction> => {
+    const status: GenSaleStatus = await getSaleStatus(contract);
+
+    errorCheck(
+        status === GenSaleStatus.NotStarted,
+        "Cannot purchase a domain when sale has not started"
+    );
+
+    errorCheck(status === GenSaleStatus.Ended, "Sale has already ended");
+
+    errorCheck(count.eq("0"), "Cannot purchase 0 domains");
+
+    const paused = await contract.paused();
+    errorCheck(paused, "Sale contract is paused");
+
+    const domainsSold = await contract.domainsSold();
+    const numberForSale = await contract.amountForSale();
+    const newSoldNumber = domainsSold.add(count);
+    errorCheck(
```

Comment:
We shouldn't error in this case. If the request is for 5 domains, but there are only 3 in the sale we make the purchase of 3 domains and refund any extra ETH